src/inmanta/plugins.py

```diff
@@ -191,21 +192,160 @@ def clear(cls, inmanta_module: Optional[str] = None) -> None:
+        self.type_expression = type_expression
+
+        # We define the attribute but don't set it yet, this will be done when
+        # the type is resolved.  This allows to differentiate between a type that
+        # has not been resolved and a type that is a match for "any" (None).
+        self._resolved_type: Optional[inmanta_type.Type]
+
+    @property
+    def resolved_type(self) -> Optional[inmanta_type.Type]:
+        """
+        Get the resolved type of this plugin io.  The resolved type can only be accessed
+        once this object has been normalized (which happens during the plugin normalization).
+        """
+        if not hasattr(self, "_resolved_type"):
+            raise CompilerException(
+                f"{type(self).__name__} {self.IO_NAME} ({repr(self.type_expression)}) has not been normalized, "
+                "its resolved type can't be accessed."
+            )
+        return self._resolved_type
+
+    def resolve_type(self, plugin: "Plugin", resolver: Namespace) -> Optional[inmanta_type.Type]:
+        """
+        Convert the string representation of this argument's type to a type.
+        If no type annotation is present or if the type annotation allows any type to be passed
+        as argument, then None is returned.
+
+        :param plugin: The plugin that this argument is part of.
+        :param resolver: The namespace that can be used to resolve the type annotation of this
+            argument.
+        """
+        if self.type_expression is None:
+            self._resolved_type = None
+            return self._resolved_type
+
+        if not isinstance(self.type_expression, str):
+            raise CompilerException(
+                "Bad annotation in plugin %s for %s, expected str but got %s (%s)"
+                % (plugin.get_full_name(), self.IO_NAME, type(self.type_expression).__name__, self.type_expression)
+            )
+
+        plugin_line: Range = Range(plugin.location.file, plugin.location.lnr, 1, plugin.location.lnr + 1, 1)
+        locatable_type: LocatableString = LocatableString(self.type_expression, plugin_line, 0, resolver)
+        self._resolved_type = resolve_type(locatable_type, resolver)
+        return self._resolved_type
+
+    def validate(self, value: object) -> bool:
+        """
+        Validate that the given value can be passed to this argument.  Returns True if the value is known
+        and valid, False if the value is unknown, and raises a ValueError is the value is not of the
+        expected type.
+
+        :param value: The value to validate
+        """
+        if isinstance(value, Unknown):
+            # Value is not known, it can not be validated
+            return False
+
+        if self.resolved_type is None:
+            # Any value is valid
+            return True
+
+        # Validate the value, use custom validate method of the type if it exists
+        valid = self.resolved_type.validate(value)
```

Comment:
Code-wise, this is a breaking change, before, it was equivalent to this:
```python
        if hasattr(self.resolved_type, "validate"):
            valid = getattr(self.resolved_type, "validate")(value)
        else:
            valid = isinstance(value, self.resolved_type)
```
But typing-wise, it makes no sense, as the value in `resolved_type` comes from `Namespace.get_type` and that one only returns an inmanta `Type` object.  Is it okay to drop it then? Or are we not confident enough that the typing is correct?